baseplate/sidecars/live_data_watcher.py

```diff
@@ -128,25 +73,67 @@ def on_change(self, data: Optional[bytes], _znode_stat: ZnodeStat) -> None:
             if self.owner and self.group:
                 os.fchown(tmpfile.fileno(), self.owner, self.group)
             os.fchmod(tmpfile.fileno(), self.mode)
-            try:
-                json_data = json.loads(data.decode("UTF-8"))
-            except json.decoder.JSONDecodeError:
-                # If JSON fails to decode, still write the bytes data since
-                # we don't necessarily know if the the contents of the znode
-                # had to be valid JSON anyways.
-                tmpfile.write(data)
-            else:
-                # If no exceptions, we have valid JSON, and can parse accordingly.
-                data = NodeWatcher.get_data_to_write(json_data, data)
-                if data is None:
-                    logger.warning(
-                        "No data written to destination file. Something is likely misconfigured."
-                    )
-                    return
-                tmpfile.write(data)
+            tmpfile.write(data)
         os.rename(self.dest + ".tmp", self.dest)
 
 
+def _parse_loader_type(data: Optional[bytes]) -> LoaderType:
+    try:
+        json_data = json.loads(data.decode("UTF-8"))
+    except json.decoder.JSONDecodeError:
+        logger.debug("Data is not valid JSON. Loading as RAW")
+        return LoaderType.RAW
+
+    try:
+        loader_type = json_data["live_data_watcher_load_type"]
+    except (json.decoder.JSONDecodeError, KeyError, AttributeError, TypeError):
+        logger.debug("Data is not a JSON object like {\"live_data_watcher_load_type\":\"load_type\"}. Loading as RAW")
+        return LoaderType.RAW
+
+    return LoaderType(loader_type)
+
+
+def _load_from_s3(data: bytes) -> Optional[bytes]:
+    loader_config = json.loads(data.decode("UTF-8"))
+    try:
+        region_name = loader_config["region_name"]
+        s3_kwargs = {
+            "Bucket": loader_config["bucket_name"],
+            "Key": loader_config["file_key"],
+            "SSECustomerKey": loader_config["sse_key"],
+            "SSECustomerAlgorithm": "AES256",
+        }
+    except KeyError as e:
+        # We require all of these keys to properly read from S3.
+        logger.exception(
+                "Failed to update live config: unable to fetch content from s3: source config has invalid or missing keys: %s.",
+                e.args[0]
+        )
+        return None
+
+    s3_client = boto3.client(
+        "s3",
+        region_name=region_name,
+    )
+
+    try:
+        s3_object = s3_client.get_object(**s3_kwargs)
+        # Returns bytes.
+        return s3_object["Body"].read()
+    except ClientError as error:
+        logger.exception(
+            "Failed to update live config: failed to load data from S3. Received error code %s: %s",
+            error.response["Error"]["Code"],
+            error.response["Error"]["Message"],
+        )
+    except ValueError as error:
+        logger.exception(
```

Comment:
🔕  I think technically it's more pythonic to just raise an exception than to log it and return `None`. Doesn't make much of a difference to me.

Reply:
If we raise an exception then we either have to handle it elsewhere or have the data watcher fail and exit.

IMO returning None means "we got no data" and that's a valid thing to write out as data.

I can be convinced otherwise pretty easily

Reply:
"No data" valid aka empty file (though I'd expect an empty file locally too)

We tried to get this and got an error (4xx, 5xx), I would not expect that to dump an empty file

Reply:
If it were to return None, you'd just end up getting an exception when trying to write the file instead:

https://github.com/reddit/baseplate.py/pull/676/files#diff-f08bba0851fd42e4ae2a8d9fb6b983e377e46214ca4fbb35a3738f4d97665630R76

![Screen Shot 2022-05-20 at 2 28 50 PM](https://user-images.githubusercontent.com/1282393/169590519-2de3bcf1-6ef7-4280-9e06-00beb4163322.png)

Reply:
hm ok I can see it. Another edge case to consider then:

Let's say the s3 object was created, we had watchers fetch it, then we deleted it from s3 and fetchers ran again. Now they've got an "object doesn't exist error" (or whatever the error is). Do we delete the existing local copy? Leave it alone but with old data?

Reply:
I'd opt to leave it alone but make that error loud A F. We don't want to bring down a service because someone accidentally deleted something, but we also don't want that to go silently unnoticed.

Reply:
oh dang we handled this case in the previous form and I accidentally ripped it out, good catch